src/classlibnative/bcltype/bignum.h

```diff
@@ -0,0 +1,152 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+//
+// File: bignum.h
+//
+
+//
+
+#ifndef _BIGNUM_H_
+#define _BIGNUM_H_
+
+#include <clrtypes.h>
+
+class BigNum
+{
+public:
+    BigNum();
+    BigNum(UINT32 value);
+    BigNum(UINT64 value);
+    ~BigNum();
+
+    BigNum & operator=(const BigNum &rhs);
+
+    static UINT32 LogBase2(UINT32 val);
+    static UINT32 LogBase2(UINT64 val);
+
+    static int Compare(const BigNum& lhs, const BigNum& rhs);
+
+    static void ShiftLeft(UINT64 input, UINT32 shift, BigNum& output);
+    static void Pow10(int exp, BigNum& result);
+    static void PrepareHeuristicDivide(BigNum* pDividend, BigNum* divisor);
+    static UINT32 HeuristicDivide(BigNum* pDividend, const BigNum& divisor);
+    static void Multiply(const BigNum& lhs, UINT32 value, BigNum& result);
+    static void Multiply(const BigNum& lhs, const BigNum& rhs, BigNum& result);
+
+    bool IsZero() const;
+
+    void Multiply(UINT32 value);
+    void Multiply(const BigNum& value);
+    void Multiply10();
+    void ShiftLeft(UINT32 shift);
+    void SetUInt32(UINT32 value);
+    void SetUInt64(UINT64 value);
+    void SetZero();
+    void ExtendBlock(UINT32 blockValue);
+    void ExtendBlocks(UINT32 blockValue, UINT32 blockCount);
+
+private:
+
+    static const UINT32 BIGSIZE = 35;
+    static const UINT32 UINT32POWER10NUM = 8;
+    static const UINT32 BIGPOWER10NUM = 6;
+    static constexpr UINT32 m_power10UInt32Table[UINT32POWER10NUM] = 
```

Comment:
It doesn't need to be addressed now, but I think it would be worth investigating if holding up to 10^9 in here and then incrementing `m_power10BigNumTable` by 10, rather than 8 would be more efficient (I think it should result in fewer overall lookups).